src/sentry/api/serializers/models/rule.py

```diff
@@ -289,3 +302,192 @@ def serialize(self, obj, attrs, user, **kwargs) -> RuleSerializerResponse:
             d["disableDate"] = attrs["disable_date"]
 
         return d
+
+
+class WorkflowEngineRuleSerializer(Serializer):
+    def __init__(
+        self,
+        expand: list[str] | None = None,
+        prepare_component_fields: bool = False,
+        project_slug: str | None = None,
+    ):
+        super().__init__()
+        self.expand = expand or []
+        self.prepare_component_fields = prepare_component_fields
+        self.project_slug = project_slug
+
+    def get_attrs(self, item_list: Sequence[Workflow], user, **kwargs):
+        from sentry.workflow_engine.migration_helpers.rule_conditions import (
+            translate_to_rule_condition_filters,
+        )
+
+        # Bulk fetch environments
+        environments = Environment.objects.in_bulk(
+            [_f for _f in [i.environment_id for i in item_list] if _f]
+        )
+
+        # Bulk fetch users that created workflows
+        users = {
+            u.id: u
+            for u in user_service.get_many_by_id(
+                ids=[item.created_by_id for item in item_list if item.created_by_id is not None]
+            )
+        }
+
+        # Bulk fetch projects for workflows (attached through detectors)
+        workflow_to_projects: dict[Workflow, set[Project]] = defaultdict(set)
+        detector_workflows = DetectorWorkflow.objects.filter(
+            workflow_id__in=[item.id for item in item_list]
+        ).prefetch_related("detector__project")
+        for detector_workflow in detector_workflows:
+            workflow_to_projects[detector_workflow.workflow].add(detector_workflow.detector.project)
+
+        # Bulk fetch trigger and filter conditions
+        # Get trigger condiions
+        workflows = (
+            Workflow.objects.filter(id__in=[wf.id for wf in item_list])
+            .select_related("when_condition_group")
+            .prefetch_related("when_condition_group__conditions")
+        )
+
+        # Get filter conditions
+        workflow_dcg_prefetch = Prefetch(
+            "workflowdataconditiongroup_set",
+            queryset=WorkflowDataConditionGroup.objects.prefetch_related(
+                "condition_group__conditions"
+            ),
+            to_attr="prefetched_wdcgs",
+        )
+        workflows_and_filters = Workflow.objects.filter(
+            id__in=[wf.id for wf in item_list]
+        ).prefetch_related(workflow_dcg_prefetch)
+
+        workflow_to_filters = {}
+        workflow_to_filter_match = {}
+        for wf in workflows_and_filters:
+            # pick first DCG (rules only have 1)
+            wdcg = wf.prefetched_wdcgs[0]  # type: ignore[attr-defined]
```

Comment:
should this just use `.first()`? curious about the type ignore too -- would .first fix that?

Reply:
it thinks it doesn't exist because the attribute is attached at runtime via the Prefetch, and mypy doesn't know it exists

Reply:
somehow it seems to be ok removing it now

Reply:
jk, adding the type back makes it not think this is an attribute